detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryFilter.kt

```diff
@@ -0,0 +1,70 @@
+package io.gitlab.arturbosch.detekt.rules.style
+
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Debt
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Severity
+import io.gitlab.arturbosch.detekt.api.CodeSmell
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtLambdaExpression
+import org.jetbrains.kotlin.psi.psiUtil.nextLeaf
+import org.jetbrains.kotlin.psi.unpackFunctionLiteral
+
+/**
+ * Unnecessary filter add complexity to the code and accomplish noting. They should be removed.
+ *
+ * <noncompliant>
+ * val x = listOf(1, 2, 3)
+ *      .filter { it > 1 }
+ *      .count()
+ *
+ * val x = listOf(1, 2, 3)
+ *      .filter { it > 1 }
+ *      .isEmpty()
+ * </noncompliant>
+ *
+ * <compliant>
+ * val x = listOf(1, 2, 3)
+ *      .count { it > 2 }
+ * }
+ *
+ * val x = listOf(1, 2, 3)
+ *      .none { it > 1 }
+ *
+ * </compliant>
+ */
+class UnnecessaryFilter(config: Config = Config.empty) : Rule(config) {
+    override val issue: Issue = Issue("UnnecessaryFilter", Severity.Style,
+        "filter() with other collection operations may be simplified.",
+        Debt.FIVE_MINS)
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        super.visitCallExpression(expression)
+
+        val calleeExpression = expression.calleeExpression
+        if (calleeExpression?.text != "filter") return
+
+        expression.checkNextLeaf("size")
+        expression.checkNextLeaf("count")
+        expression.checkNextLeaf(
+            "isEmpty", "any")
+        expression.checkNextLeaf(
+            "isNotEmpty", "none")
```

Comment:
This will yield false-positives, if you have property accessors like this. It's not checked, whether the calling function is actually a high order function named `filter` with a lambda expression in it.

```
call().filter.size
```

```suggestion
        expression.checkNextLeaf("isEmpty", "any")
        expression.checkNextLeaf("isNotEmpty", "none")
```

Reply:
> whether the calling function is actually a high order function named filter

nit: To be 100% exact we should also verify that the calling method is the "filter" from the Kotlin stdlib (but that will require type resolution)

Reply:
Ah that's why you were suggesting adding requiresTypeResolution in the first place.